cmd/lotus-chainwatch/processor/processor.go

```diff
@@ -0,0 +1,301 @@
+
+	"github.com/filecoin-project/lotus/api"
+	"github.com/filecoin-project/lotus/chain/types"
+	"github.com/filecoin-project/lotus/lib/parmap"
+)
+
+var log = logging.Logger("processor")
+
+type Processor struct {
+	db *sql.DB
+
+	node api.FullNode
+
+	// number of blocks processed at a time
+	batch int
+}
+
+type ActorTips map[types.TipSetKey][]actorInfo
+
+type actorInfo struct {
+	act types.Actor
+
+	stateroot cid.Cid
+	height    abi.ChainEpoch // so that we can walk the actor changes in chronological order.
+
+	tsKey       types.TipSetKey
+	parentTsKey types.TipSetKey
+
+	addr  address.Address
+	state string
+}
+
+func NewProcessor(db *sql.DB, node api.FullNode, batch int) *Processor {
+	return &Processor{
+		db:    db,
+		node:  node,
+		batch: batch,
+	}
+}
+
+func (p *Processor) setup() error {
+	if err := p.setupMiners(); err != nil {
+		return err
+	}
+
+	if err := p.setupRewards(); err != nil {
+		return err
+	}
+
+	if err := p.setupMessages(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (p *Processor) Start(ctx context.Context) {
+	log.Info("Starting Processor")
+
+	if err := p.setup(); err != nil {
+		log.Fatalw("Failed to setup processor", "error", err)
+	}
+
+	go p.subMpool(ctx)
+
+	// main processor loop
+	go func() {
+		for {
+			toProcess, err := p.unprocessedBlocks(ctx, p.batch)
+			if err != nil {
+				log.Fatalw("Failed to get unprocessed blocks", "error", err)
+			}
+
+			// TODO special case genesis state handling here to avoid all the special cases that will be needed for it else where
+			// before doing "normal" processing.
+
+			actorChanges, err := p.collectActorChanges(ctx, toProcess)
+			if err != nil {
+				log.Fatalw("Failed to collect actor changes", "error", err)
+			}
+
+			grp, ctx := errgroup.WithContext(ctx)
+
+			grp.Go(func() error {
+				if err := p.HandleMinerChanges(ctx, actorChanges[builtin.StorageMinerActorCodeID]); err != nil {
+					return xerrors.Errorf("Failed to handle miner changes: %w", err)
+				}
+				return nil
+			})
+
+			grp.Go(func() error {
+				if err := p.HandleRewardChanges(ctx, actorChanges[builtin.RewardActorCodeID]); err != nil {
+					return xerrors.Errorf("Failed to handle reward changes: %w", err)
+				}
+				return nil
+			})
+
+			grp.Go(func() error {
+				if err := p.HandleMessageChanges(ctx, toProcess); err != nil {
+					return xerrors.Errorf("Failed to handle message changes: %w", err)
+				}
+				return nil
+			})
+
+			if err := grp.Wait(); err != nil {
+				log.Errorw("Failed to handle actor changes...retrying", "error", err)
+				continue
+			}
+
+			if err := p.markBlocksProcessed(ctx, toProcess); err != nil {
+				log.Fatalw("Failed to mark blocks as processed", "error", err)
+			}
+
+			if err := p.refreshViews(); err != nil {
+				log.Errorw("Failed to refresh views", "error", err)
+			}
+		}
+	}()
+
+}
+
+func (p *Processor) refreshViews() error {
+	if _, err := p.db.Exec(`refresh materialized view state_heights`); err != nil {
```

Comment:
Let's run this with `CONCURRENTLY` which leaves the table available for query while it's updating. This requires that we add a `UNIQUE` index on `state_heights` which can be on the `parentstateroot` column.
```suggestion
	if _, err := p.db.Exec(`refresh materialized view concurrently state_heights`); err != nil {
```